hal/armv7m/stm32/l4/stm32l4.c

```diff
@@ -283,22 +335,36 @@ int _stm32_rccSetCPUClock(u32 hz)
 	}
 */
 	else {
-		/* We can use HSI, if higher frequency is needed */
+		/* Not supported */
 		return -EINVAL;
 	}
 
 	if (hz > 6000 * 1000)
 		_stm32_pwrSetCPUVolt(1);
 
-	while (!(*(stm32_common.rcc + rcc_cr) & 2))
-		;
+	if (hz == 16 * 1000 * 1000) {
+		/* We can use HSI */
+		_stm32_rccMsiToHsi();
 
-	t = *(stm32_common.rcc + rcc_cr) & ~(0xf << 4);
-	*(stm32_common.rcc + rcc_cr) = t | range << 4 | (1 << 3);
-	hal_cpuDataMemoryBarrier();
+		/* Use HSI after STOP2 wakeup */
+		*(stm32_common.rcc + rcc_cfgr) |= 1 << 15;
+	}
+	else {
+		while (!(*(stm32_common.rcc + rcc_cr) & 2))
+			;
 
-	if (hz <= 6000 * 1000)
-		_stm32_pwrSetCPUVolt(2);
+		t = *(stm32_common.rcc + rcc_cr) & ~(0xf << 4);
+		*(stm32_common.rcc + rcc_cr) = t | range << 4 | (1 << 3);
+		hal_cpuDataMemoryBarrier();
+
+		if (hz <= 6000 * 1000)
+			_stm32_pwrSetCPUVolt(2);
```

Comment:
This should be done after switching to MSI to avoid violating wait states requirements

Reply:
Right, will fix